bootstore/src/schemes/v0/state.rs

```diff
@@ -0,0 +1,594 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! State for the V0 protocol state machine
+
+use super::fsm::StateHandler;
+use super::fsm_output::{ApiError, ApiOutput};
+use super::messages::{Envelope, Request, RequestType};
+use super::state_learned::LearnedState;
+use super::state_learning::LearningState;
+use super::state_uninitialized::UninitializedState;
+use super::{fsm_output::Output, state_initial_member::InitialMemberState};
+use crate::{
+    trust_quorum::{LearnedSharePkgV0, RackSecret, SharePkgV0},
+    Sha3_256Digest,
+};
+use derive_more::From;
+use serde::{Deserialize, Serialize};
+use sha3::{Digest, Sha3_256};
+use sled_hardware::Baseboard;
+use std::collections::{BTreeMap, BTreeSet};
+use std::fmt::Debug;
+use uuid::Uuid;
+use zeroize::{Zeroize, ZeroizeOnDrop};
+
+/// A number of clock ticks from some unknown epoch
+pub type Ticks = usize;
+
+// An index into an encrypted share
+#[derive(
+    Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize,
+)]
+pub struct ShareIdx(pub usize);
+
+/// Configuration of the FSM
+#[derive(Debug, Clone, Copy)]
+pub struct Config {
+    pub learn_timeout: Ticks,
+    pub rack_init_timeout: Ticks,
+    pub rack_secret_request_timeout: Ticks,
+}
+
+/// In memory state shared by all 4 FSM states
+#[derive(Debug)]
+pub struct FsmCommonData {
+    /// Unique IDs of this peer
+    pub id: Baseboard,
+
+    /// Configuration of the FSM
+    pub config: Config,
+
+    /// Unique IDs of connected peers
+    pub peers: BTreeSet<Baseboard>,
+
+    /// The current time in ticks since the creation of the FSM
+    pub clock: Ticks,
+
+    /// Any shares we are in the process of retrieving or a computed rack
+    /// secret. If no collection is ongoing and we are not holding a computed
+    /// rack secret, then the state is `Empty`. The state is always empty for
+    /// `State::Uninitialized` and `State::Learning`
+    pub rack_secret_state: RackSecretState,
+}
+
+impl FsmCommonData {
+    pub fn new(id: Baseboard, config: Config) -> FsmCommonData {
+        FsmCommonData {
+            id,
+            config,
+            peers: BTreeSet::new(),
+            clock: 0,
+            rack_secret_state: RackSecretState::Empty,
+        }
+    }
+
+    // Round-robin peer selection
+    pub fn next_peer(&self, current: &Baseboard) -> Option<Baseboard> {
+        if let Some(index) = self.peers.iter().position(|x| x == current) {
+            let next_index = (index + 1) % self.peers.len();
+            self.peers.iter().nth(next_index).cloned()
+        } else {
+            self.peers.first().cloned()
+        }
```

Comment:
This is O(N), because it's iterating over elements in the set.  You could get it down to `O(log(N))` by using range operations instead:
```
        let mut iter = self.peers.range(current..);
        let matched = iter.next() == Some(current);
        iter.next()
            .filter(|_| matched)
            .cloned()
            .or_else(|| self.peers.first().cloned())
```

Reply:
Fixed in [98ada27](https://github.com/oxidecomputer/omicron/pull/3385/commits/98ada27b53d4429463cde94cc91bedec0bc4ef98)

Thanks for the great suggestion!